pkg/mapreconciler/reconciler.go

```diff
@@ -0,0 +1,276 @@
+		m:          m,
+		db:         db,
+		tbl:        tbl,
+		pkIndex:    pkIndex,
+		jobGroup:   jobGroup,
+		options:    options,
+		retryQueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+}
+
+type mapReconciler[E Pair[K, V], K Marshalable, V any] struct {
+	m          Map[K, V]
+	db         *statedb.DB
+	tbl        statedb.Table[E]
+	pkIndex    statedb.Index[E, K]
+	jobGroup   job.Group
+	retryQueue workqueue.RateLimitingInterface
+	options    options[E, K, V]
+}
+
+// Start implements hive.Hook.
+func (m *mapReconciler[E, K, V]) Start(ctx hive.HookContext) error {
+	// Only do work if the map is enabled.
+	if m.m.Enabled() {
+		m.jobGroup.Add(
+			job.OneShot(fmt.Sprintf("map-%s-reconciler", m.m.Name()), m.reconcile),
+			job.OneShot(fmt.Sprintf("map-%s-retirer", m.m.Name()), m.Retirer),
+		)
+	}
+
+	return nil
+}
+
+// Start implements hive.Hook.
+func (m *mapReconciler[E, K, V]) Stop(ctx hive.HookContext) error {
+	m.retryQueue.ShutDown()
+	return nil
+}
+
+func (m *mapReconciler[E, K, V]) newDeleteTracker() (*statedb.DeleteTracker[E], error) {
+	txn := m.db.WriteTxn(m.tbl)
+	tracker, err := m.tbl.DeleteTracker(txn, fmt.Sprintf("map-%s", m.m.Name()))
+	if err != nil {
+		txn.Abort()
+		return nil, fmt.Errorf("delete tracker: %w", err)
+	}
+	txn.Commit()
+
+	return tracker, nil
+}
+
+// reconcile is the main reconciliation loop for the map reconciler.
+func (m *mapReconciler[E, K, V]) reconcile(ctx context.Context) error {
+	tracker, err := m.newDeleteTracker()
+	if err != nil {
+		return err
+	}
+
+	// On startup always perform a full reconciliation so we always start from a known state.
+	rev := m.fullReconciliation(ctx)
+
+	limitTicker := time.NewTicker(m.options.partialReconcileRatelimit)
+	fullReconcileTicker := time.NewTicker(m.options.fullReconciliationInterval)
+
+	for {
+		// limit the reconciliation rate, better to batch changes than to loop
+		// for every individual change.
+		<-limitTicker.C
+
+		var invalidate <-chan struct{}
+		rev, invalidate = m.partialReconciliation(ctx, tracker, rev)
+
+		select {
+		case <-invalidate:
+			continue
+
+		case <-fullReconcileTicker.C:
+			rev = m.fullReconciliation(ctx)
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// partialReconciliation processes the changes in the delete tracker and applies them to the BPF map.
+// It doesn't reconcile values that have not changed and doesn't check the map for rouge values, this makes it fast.
+func (m *mapReconciler[E, K, V]) partialReconciliation(ctx context.Context, tracker *statedb.DeleteTracker[E], revIn uint64) (uint64, <-chan struct{}) {
+	rxn := m.db.ReadTxn()
+	rev, invalidate, err := tracker.Process(rxn, revIn, func(entry E, deleted bool, _ uint64) error {
+		if deleted {
+			err := m.m.Delete(entry.Key())
+			// If we error, retry this key later, don't hold up the delete tracker.
+			if err != nil && !errors.Is(err, ebpf.ErrKeyNotExist) {
+				m.retryQueue.AddRateLimited(entry.Key())
```

Comment:
Doesn't this lead to races? E.g. it might be we'll retry the delete while it was added back? `Forget` doesn't remove anything from the queue.

Reply:
The idea is that we only store the key for retries, just like Resource[T] so it doesn't matter if something happened in the table between us adding the key and the actual retry action. Not sure this warrants a custom retry queue, although it might make testing and verification easier :thinking:

Reply:
It's more that the snapshot that the retry gets might be older or newer than what we have here, e.g. that's why it might travel back in time.